### ISSUE TITLE
libcurl 8.5.0 breaks mpd http streaming

### ISSUE THREAD

**Issue by bubbleguuum** [curl/curl#12632]
### I did this

libcurl  v8.5.0 is causing issues in mpd http streaming, for http streams whose size is > 128 MB (which causes mpd to not fully buffer the stream in memory). Shortly after playing such stream, a curl API inexplicably returns CURLE_OUT_OF_MEMORY, causing mpd to immediately stop playback. libcurl v8.4.0 and prior work just fine.

I reported all the details and findings about it and how to reproduce it on this [mpd issue](https://github.com/MusicPlayerDaemon/MPD/issues/1952).

The problematic commit has been found to be https://github.com/curl/curl/commit/47f5b1a37f3e8055dfea446236c05478087858dc via bissecting.

After analysis, the mpd developer thinks this is a curl issue, hence this report.

Interestingly, mpd running with git libcurl (as of the date of this post) do not cause an OOM but instead will make the mpd "io" thread (making curl API calls) to take 100% CPU.

Any insights for fixing this issue would be appreciated.

### curl/libcurl version

libcurl 8.5.0

### operating system

openSUSE Tumbleed (this issue is not OS dependent).

**Comment by bagder** [comment]
Pointing to a cryptic issue in another application is not helping us much. Please describe a lot more about the actual problem and how we can reproduce it - without running or debugging someone else's complicated application.

**Comment by bubbleguuum** [comment]
mpd developer: not my problem
curl developer: not my problem

After spending several hours on this issue to narrow it down and writing reports, I'm sick of it. Feel free to close.
Let's keep mpd crashing.

**Comment by bagder** [comment]
A problem with curl *is* a problem for us, but for us to be able to work on a reported issue we need some guidance - a detailed description to start with. If neither you nor any other mpd user/developer want to help with that, then I cannot see how we can act on this report. Then it will be closed, yes. We are not mind-readers, unfortunately.

**Comment by bubbleguuum** [comment]
If you want more details, ask mpd developer MaxKellerman, assuming he his willing to help, which I doubt.
I've done my part at this point.

**Comment by bagder** [comment]
Sorry, but if nobody can describe **a curl problem** to us then there is no curl problem as far as I am concerned.

**Comment by bubbleguuum** [comment]
Did you read this analysis from the mpd developer wit the change in behavior between 6.4.0 and 6.5.0:

https://github.com/MusicPlayerDaemon/MPD/issues/1952#issuecomment-1875905999

**Comment by bagder** [comment]
That's a description of mpd behavior when it uses libcurl together with some strace dumps. I have exactly *zero* knowledge of what mpd does with libcurl or how it works in general so I cannot tell what most of that means in terms of curl problems or not. What APIs does it use? What protocol (version) does it speak? What exactly is mpd doing when this issue is triggered? Which libcurl function returns what error?

**Comment by bubbleguuum** [comment]
Well, I'll leave it to developers of 2 large software to figure it out (eventually) by communicating. I'm expert in neither the curl nor mpd codebase.

**Comment by jay** [comment]
@MaxKellermann

**Comment by MaxKellermann** [comment]
MPD uses the CURL "multi" API and uses `CURLMOPT_SOCKETFUNCTION` to integrate it in its epoll-based I/O event loop. See https://github.com/MusicPlayerDaemon/MPD/blob/master/src/lib/curl/Global.cxx for the part that does this integration. There is nothing special here, I guess every program using CURL does it that way, that's how CURL is supposed to be integrated in non-blocking I/O loops, is it not?

So the problem is that returning `CURL_WRITEFUNC_PAUSE` used to lead to a `CURLMOPT_SOCKETFUNCTION` call that would unregister the HTTP connection socket, effectively pausing the transfer (in CURL 8.4.0).
That `CURLMOPT_SOCKETFUNCTION` is no longer performance by CURL 8.5.0, and instead, CURL keeps receiving from the "paused" socket, buffering received data until the buffer eventually grows too large, causing `dyn_nappend()` to return `CURLE_OUT_OF_MEMORY`, failing the whole HTTP request.

I expect `CURL_WRITEFUNC_PAUSE` to effectively pause the transfer and never grow the `dynbuf` too large.

CURL 8.5.0 does not actually pause the transfer and not receive any data; but all it actually does since 8.5.0 is omit the `WRITEFUNCTION` calls. But it really keeps receiving+buffering data.

This is a CURL regression; at best, this is a semantic ABI change which breaks at least one widely used application after the update. But I believe this is really a bug, not just some minor semantic change.

**Comment by bagder** [comment]
@MaxKellermann HTTP/1 or HTTP/2 ?

**Comment by MaxKellermann** [comment]
I only tested with 1.

**Comment by MaxKellermann** [comment]
@icing with your patch on top of 8.5.0, the problem is gone and MPD can play the song.

**Comment by bubbleguuum** [comment]
And I confirm patch works with libcurl master as well (without it, it causes 100% CPU usage in the mpd io thread).

**Comment by ClaudioSantacesaria** [comment]
Hi,
can I ask for help here? I still have the problem with curl 8.6.0. I have same configuration (using BubbleUpnp Android with Qobuz, upnpdcli and MPD) and playback interrupts after 0-15 seconds when file is > 128 MByte.
I tried:
```
mpc clear
mpc add  http://bubblesoftapps.com/tmp/sine_crash.flac
mpc play
```
plays 12 seconds and then stops. MPD logs reports
```
Feb 08 23:02 : exception: CURL failed: Out of memory
Feb 08 23:02 : player: played "http://bubblesoftapps.com/tmp/sine_crash.flac"
```

CURL is 8.6.0
```
# curl -V
curl 8.6.0 (x86_64-pc-linux-gnu) libcurl/8.6.0 OpenSSL/3.2.1 zlib/1.3.1 brotli/1.1.0 zstd/1.5.5 libidn2/2.3.7 libpsl/0.21.2 libssh2/1.11.0 nghttp2/1.59.0 nghttp3/1.1.0
Release-Date: 2024-01-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```
MPD is 0.23.15
```
# mpd -V
Music Player Daemon 0.23.15 (0.23.15)
```
Any suggestion. Can this be double checked if it still works in the released 8.6.0?
